File: frontend/src/lib/utils/sanitization.ts
import DOMPurify from 'isomorphic-dompurify';

/**
 * Sanitize user-generated HTML content safely.
 * Works on both client-side and server-side (isomorphic).
 */
export const sanitizeHTML = (html: string): string => {
    return DOMPurify.sanitize(html, {
        ALLOWED_TAGS: [
            'b', 'i', 'em', 'strong', 'a', 'p', 'br', 'ul', 'ol', 'li',
            'blockquote', 'code', 'pre', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
            'span', 'div', 'img'
        ],
        ALLOWED_ATTR: ['href', 'target', 'src', 'alt', 'class', 'title'],
    });
};

/**
 * Strictly sanitize content, removing all HTML tags.
 */
export const sanitizeStrict = (text: string): string => {
    return DOMPurify.sanitize(text, { ALLOWED_TAGS: [] });
};

File: backend/internal/utils/sanitization.go
package utils

import (
	"github.com/microcosm-cc/bluemonday"
)

var (
	// StrictPolicy removes all HTML tags
	StrictPolicy = bluemonday.StrictPolicy()

	// UGCPolicy allows basic formatting for user-generated content
	UGCPolicy = bluemonday.UGCPolicy()
)

// SanitizeHTML removes dangerous HTML tags and attributes
func SanitizeHTML(input string) string {
	return UGCPolicy.Sanitize(input)
}

// SanitizeStrict removes ALL HTML tags
func SanitizeStrict(input string) string {
	return StrictPolicy.Sanitize(input)
}
